File: BIT - Bachelor IT/5.term/IW5 - Programming .NET and C#/Tournament.API.BL/Facades/FacadeBase.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using AutoMapper;
using Tournament.API.BL.Facades.Interfaces;
using Tournament.API.BL.Models.DetailModels.Interfaces;
using Tournament.API.BL.Models.FormDefaultModels.Interfaces;
using Tournament.API.BL.Models.FormModels.Interfaces;
using Tournament.API.BL.Models.ListModels.Interfaces;
using Tournament.API.BL.Models.SelectModels.Interfaces;
using Tournament.API.Common.Interfaces;
using Tournament.API.DAL.Entities.Interfaces;
using Tournament.API.DAL.Repositories.Interfaces;
using Tournament.API.DAL.UnitOfWork.Interfaces;
using Microsoft.AspNetCore.Hosting;
using Tournament.API.Common.Enums;
using Tournament.API.DAL.UnitOfWork.Factories.Interfaces;

namespace Tournament.API.BL.Facades
{
    public class FacadeBase<TDetailModel, TListModel, TFormModel, TFormDefaultModel, TEntity> : IFacadeBase<TDetailModel, TListModel, TFormModel, TFormDefaultModel>
        where TDetailModel : class, IDetailModel
        where TListModel : class, IListModel
        where TFormModel : class, IFormModel
        where TFormDefaultModel : class, IFormDefaultModel, new()
        where TEntity : class, IEntity
    {
        protected readonly IUnitOfWork UnitOfWork;
        protected readonly IRepository<TEntity> Repository;
        protected readonly IMapper Mapper;
        protected readonly Entities Entity;
        

        protected virtual TListModel GetListModel(TListModel listModel)
        {
            return listModel;
        }

        protected virtual IList<TListModel> GetListModels(IList<TListModel> listModels)
        {
            foreach (var listModel in listModels)
            {
                GetListModel(listModel);
            }

            return listModels;
        }
        

        public FacadeBase(IUnitOfWork unitOfWork, IMapper mapper, Entities entity)
        {
            UnitOfWork = unitOfWork;
            Entity = entity;
            Repository = UnitOfWork.GetRepository<TEntity>(Entity);
            this.Mapper = mapper;
        }

        public IList<TListModel> GetAll()
        {
            var entities = Repository.GetAll();
            var listModels = Mapper.Map<IList<TListModel>>(entities);
            return GetListModels(listModels);
        }

        public TDetailModel? GetById(Guid id)
        {
            return Mapper.Map<TDetailModel?>(Repository.GetById(id));
        }

        public virtual TListModel Add(TFormModel formModel)
        {
            var detailModel = Mapper.Map<TListModel>(Repository.Insert(Mapper.Map<TEntity>(formModel)));
            UnitOfWork.Commit();
            return detailModel;
        }

        public virtual TListModel Update(TFormModel formModel)
        {
            var detailModel = Mapper.Map<TListModel>(Repository.Update(Mapper.Map<TEntity>(formModel)));
            UnitOfWork.Commit();
            return detailModel;
        }

        public virtual void Remove(Guid id)
        {
            Repository.Delete(id);
            UnitOfWork.Commit();
        }

        public virtual TFormDefaultModel GetDefaultFormModel()
        {
            var formDefaultModel = new TFormDefaultModel();
            return formDefaultModel;
        }

        public void Dispose()
        {
        }
    }
}

File: MIT - Master IT/2.term/PIS - Advanced Information Systems/api/src/main/java/cz/vutbr/fit/api/bl/facades/FacadeBase.java
package cz.vutbr.fit.api.bl.facades;

import cz.vutbr.fit.api.bl.facades.interfaces.IFacade;
import cz.vutbr.fit.api.bl.mappers.interfaces.IMapper;
import cz.vutbr.fit.api.bl.models.detailModels.IDetailModel;
import cz.vutbr.fit.api.bl.models.formModels.IFormModel;
import cz.vutbr.fit.api.bl.models.formModels.RoomFormModel;
import cz.vutbr.fit.api.bl.models.listModels.IListModel;
import cz.vutbr.fit.api.dal.entity.interfaces.IEntity;
import cz.vutbr.fit.api.dal.repositories.RepositoryBase;
import cz.vutbr.fit.api.dal.repositories.interfaces.IRepository;
import cz.vutbr.fit.api.dal.unitOfWork.UnitOfWork;
import jakarta.ws.rs.core.EntityPart;

import java.util.List;


public abstract class FacadeBase<TEntity extends IEntity> implements IFacade<TEntity>, AutoCloseable {

    protected RepositoryBase _repository;
    private IMapper mapper;

    protected UnitOfWork _unitOfWork;

    FacadeBase(RepositoryBase repository, IMapper mapper, UnitOfWork unitOfWork)
    {
        this._repository = repository;
        this.mapper=mapper;
        this._unitOfWork = unitOfWork;
    }

    FacadeBase()
    {

    }

    public List<IListModel> getAll()
    {
        return  mapper.MapEntitiesToList(_repository.getAll());
    }

    public void removeAll(){
        _unitOfWork.beginTransaction();
        _repository.removeAll();
        _unitOfWork.commit();
    }

    public int generateBasicAccounts()
    {

        _unitOfWork.beginTransaction();
        var result = _repository.generateBasicAccounts();
        _unitOfWork.commit();
        return result;

    }


    public long Create(List<EntityPart> formData) throws Exception
    {
        _unitOfWork.beginTransaction();
        var result = _repository.Insert(mapper.MapFormDataToEntity(formData));
        _unitOfWork.commit();
        return result;
    }

    public long Create(IEntity entity)
    {
        String msg ="";
        try{
            _unitOfWork.beginTransaction();
            var result = _repository.Insert(entity);
            _unitOfWork.commit();

            return result;
        }
        catch (Exception e) {
            msg =  e.getMessage();

        }
        return -1;
    }


    public IDetailModel getById(long id){
        return mapper.MapEntityToDetail(_repository.getById(id));
    }


    public long entityGenerator(){
        _unitOfWork.beginTransaction();
        var result = _repository.entityGenerator();
        _unitOfWork.commit();
        return result;
    }


    public long update(List<EntityPart> parts) throws Exception
    {
        _unitOfWork.beginTransaction();
        var result = _repository.Update(mapper.MapFormDataToEntity(parts));
        _unitOfWork.commit();
        return result;
    }

    public long update(IEntity entity)
    {
        _unitOfWork.beginTransaction();
        var result = _repository.Update(entity);
        _unitOfWork.commit();
        return result;
    }


    public long updateField(List<EntityPart> parts) throws Exception
    {
        var updateFieldModel = mapper.mapFormToUpdateFieldModel(parts);

        _unitOfWork.beginTransaction();
        var result = _repository.UpdateField(updateFieldModel);
        _unitOfWork.commit();
        return result;
    }



    public void removeById(long id)
    {
        _unitOfWork.beginTransaction();
        _repository.removeById(id);
        _unitOfWork.commit();
    }


    @Override
    public void close() throws Exception {
        _repository.close();
    }
}
